### PR TITLE
Added channel caching by host string

### PR DESCRIPTION
**Pull request by murgatroid99**

### CODE REVIEW

File: src/objective-c/GRPCClient/private/GRPCChannel.m [hunk 1]
```diff
@@ -40,9 +40,19 @@
...
 
+// TODO(mlumish): Investigate whether a cache with strong links is a good idea
+static NSMutableDictionary *channelCache;
+
 + (instancetype)channelToHost:(NSString *)host {
-  // TODO(jcanizales): Reuse channels.
-  return [[self alloc] initWithHost:host];
+  if (channelCache == nil) {
+    channelCache = [NSMutableDictionary dictionary];
+  }
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCChannel.m]: To be thread-safe, let's do it as in `GRPCCall.m`:102
```objective-c
+ (instancetype)channelToHost:(NSString *)host {
  static NSMutableDictionary *channelCache;
  static dispatch_once_t cacheInitialization;
  dispatch_once(&cacheInitialization, ^{
    channelCache = [NSMutableDictionary dictionary];
  });
  ...
```

File: src/objective-c/GRPCClient/private/GRPCChannel.m [hunk 2]
```diff
@@ -40,9 +40,19 @@
...
+  GRPCChannel *channel = channelCache[host];
+  if (channel == nil) {
```
Comment (jcanizales) [inline src/objective-c/GRPCClient/private/GRPCChannel.m]: nit: `if (!channel)`, for consistency with the rest of the code.

File: src/objective-c/GRPCClient/private/GRPCChannel.m (hunk 1 above)
Comment (murgatroid99) [inline src/objective-c/GRPCClient/private/GRPCChannel.m]: done

File: src/objective-c/GRPCClient/private/GRPCChannel.m (hunk 2 above)
Comment (murgatroid99) [inline src/objective-c/GRPCClient/private/GRPCChannel.m]: done

**Pull request merged**